src/uu/sort/src/sort.rs

```diff
@@ -455,71 +525,189 @@ fn compare_by(a: &str, b: &str, settings: &Settings) -> Ordering {
     };
 
     for compare_fn in &settings.compare_fns {
-        let cmp: Ordering = if settings.random {
-            random_shuffle(a, b, settings.salt.clone())
+        let cmp: Ordering = compare_fn(a, b, settings);
```

Comment:
It looks like this is only evaluating the first compare_fn, even if that one returns `Ordering::Equal`. Is that intentional?

Reply:
If I understand you, the "last resort compare" is found within the function bodies of the compare function.

But, maybe it would be best move that functionality to within the compare_by function.  If return Ordering::Equal, then call "last resort compare."

It would probably be more elegant.  I'll take a look at it.  Thanks.

Reply:
I just got confused because it is a for loop but only uses the first element :-)

Reply:
Haha, I got you.  You made me actually think, and it got me to something that is probably better.  Thanks.

Reply:
FYI, your suggestion had a performance benefit.  Our sort -n was ~20% slower than gsort, now we are about 8% faster.  We still spend way too much time on CPU because we aren't using strncmp.

hyperfine --warmup 3 'cat ~/Programming/countwords/kjvbible.txt  | ./target/release/sort -n' 'cat ~/Programming/countwords/kjvbible.txt  | gsort -n'
Benchmark #1: cat ~/Programming/countwords/kjvbible.txt  | ./target/release/sort -n
  Time (mean ± σ):      1.236 s ±  0.116 s    [User: 7.953 s, System: 0.065 s]
  Range (min … max):    1.096 s …  1.380 s    10 runs
 
Benchmark #2: cat ~/Programming/countwords/kjvbible.txt  | gsort -n
  Time (mean ± σ):      1.148 s ±  0.012 s    [User: 1.118 s, System: 0.025 s]
  Range (min … max):    1.131 s …  1.170 s    10 runs
 
Summary
  'cat ~/Programming/countwords/kjvbible.txt  | gsort -n' ran
    1.08 ± 0.10 times faster than 'cat ~/Programming/countwords/kjvbible.txt  | ./target/release/sort -n'

Reply:
FYI your comment had a performance benefit.  We used to be about ~20%+ slower than gsort because we are doing FP compares and gsort is using strncmp.  This change cuts that in half.

Reply:
I think this means that `settings.compare_fns` doesn't have to be a vector anymore, but can just be a single function pointer? (There's also no need to add `default_compare` in `uumain` to `compare_fns`)

Also, are there any performance tests in the repo? I think it would be great to add some to make sure not to regress performance (for example it looks like my PR https://github.com/uutils/coreutils/pull/1996 regresses performance quite a bit, and it would be good to have some benchmarks to optimize for)

Reply:
Agreed.  I do a hyperfine test on a bigger corpus with a few numerical opts (-n or -g) when I change something.  We might consider adding it to the tests, because occasionally I'll see it do something different than GNU.  See: https://github.com/benhoyt/countwords/blob/master/kjvbible.txt 

Performance testing would be really nice, because we are *really, really fast* at a few of these compares vs. BSD and GNU.

I think the original author's idea was to do a selected compare, then another default compare, rather than just comparing the one's which were, say, numerically equal.  The vec always had those 2 compares.  For whatever reason, the output is not the same as doing this sort of last resort compare on the numerically equal, which seems to be the POSIX way.  I've removed the for loop what you suggested and it is a bit faster.

So far we're only competitive with gsort re: wall clock time, because of par_sort_by and some inlining in my changes.  I'm sure the reason is all the FP compares.  I mean see how fast the default compare is.  We'd cook GNU with a strncmp compare.

Reply:
I'll take a stab at implementing strncmp... Assuming you haven't started doing so yet ;)

Reply:
Please!

Reply:
I PR'd your PR here: https://github.com/electricboogie/coreutils/pull/1